### PR TITLE
libobs,obs-ffmpeg: Add option for recording lossless audio

### PR DESCRIPTION
**Pull request by derrod**
### Description

Adds support for manually selecting the recording audio encoder in the UI.

Currently adds 16, 24, and 32 bit PCM as well as 24-bit ALAC and 16-bit FLAC (their respective default bit-depths in FFmpeg's order of formats).

Screenshot (note: out of date, minor changes with RFC 45, but overall still the same UX):
2023-02-18_06-19-28_xvjboa

#8498 adds the necessary UI to prevent users from selecting invalid container and audio codec combinations.

### Motivation and Context

Lossless audio recording is a highly requested feature (e.g. ideas page posts [271](https://ideas.obsproject.com/posts/271/add-flac-as-selectable-audio-codec-for-recordings), [1342](https://ideas.obsproject.com/posts/1342/record-lossless-audio), [1680](https://ideas.obsproject.com/posts/1680/allow-audio-codec-choice-on-advanced-type-standard-mode-recording-tab-only-lossless-audio), and probably many more) due to the additional flexibility in editing and avoidance of generational loss.

Additionally, PCM floating point audio allows for some additional flexibility when it comes to editing and fixing of audio in post, as clipping audio (i.e. >1.0f or <-1.0f) is retained and can be adjusted down to no longer be clipping. Aside from being able to make up for wrong recording levels in editing, this is also useful in situations where constructive interference resulted in the final output clipping while individually sources were fine (this currently happens invisibly due to the lack of a master mixer).

### How Has This Been Tested?

Recorded some videos.

### Types of changes

- New feature (non-breaking change which adds functionality)

### Checklist:

- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

### CODE REVIEW

**Comment by TianQiBuTian** [comment]
Crash on open settings.
Log: https://obsproject.com/logs/OwXxp9HwmUE0BRiz
Crash log: https://obsproject.com/logs/fkiO7swIjOTcfcJs

**Comment by derrod** [comment]
> Crash on open settings. Log: https://obsproject.com/logs/OwXxp9HwmUE0BRiz Crash log: https://obsproject.com/logs/fkiO7swIjOTcfcJs

I believe this was due to no default config entry for `AudioEncoder` being set, this should be fixed now.

**Comment by derrod** [comment]
One thing that is left to be done (ignoring conflicts with #8090) is a compatibility check.

Right now we have the following situation:
| Containers | AAC | ALAC | FLAC | Opus | PCM |
|---|---|---|---|---|---|
| FLV | ✅ | ❌ | ❌ | ❌ | ❌ |
| MP4 | ✅ | ✅ | ✅[^1] | ✅[^1] | ❌[^2] |
| MOV | ✅ | ✅ | ❌ | ❌ | ✅ |
| MKV | ✅ | ✅ | ✅ | ✅ | ✅ |
| M3U8 (HLS) | ✅ | ❌ | ❌ | ✅ | ❌ |
| TS | ✅ | ❌ | ❌ | ✅ | ❌ |

The following issues are worth noting:
- MP4 with FLAC crashes Resolve but seems to work fine in other software
- MKV support in editors remains poor
- This was not tested with #7276 

[^1]: Current obs-deps FFmpeg version requires "experimental" standards compliance (`-strict -2`), no longer necessary in latest FFmpeg
[^2]: Support is standardised, but FFmpeg doesn't support it yet. See https://trac.ffmpeg.org/ticket/10185

**Comment by derrod** [comment]
Drafting until RFC 45 stuff (especially #8090) is merged so it can be rebased and adjusted to those changes.

**Comment by alialkhatib** [comment]
I don't have anything to contribute except that I'm very excited about this. i've been recording audio separately to get 32-bit floating point audio from my audio interface (zoom f3), so this would simplify my recording setup pretty dramatically. thank you so much for doing the legwork on this

**Comment by derrod** [comment]
Updated to split out FFmpeg and PCM32 into their own commits to make them easier to skip. I'm personally not entirely sold on keeping FLAC around. Support for it in MP4 seems bad in editing software and we already have ALAC anyway.
Perhaps one of the PCM variants could also be removed, lthough compatibility is of course great there.

Also added a compatibility error message in case a non-supported combination of codec and container is chosen:
2023-03-15_03-11-52_hEJKDK

**Comment by TianQiBuTian** [comment]
Can you add `FLAC 24-bit` format?
Because FLAC supports 24-bits, and some websites need at least 24-bits to enable the `Hi-Res` option.

File: plugins/obs-ffmpeg/data/locale/en-US.ini [hunk 1]
```diff
@@ -1,6 +1,11 @@
 FFmpegOutput="FFmpeg Output"
 FFmpegAAC="FFmpeg Default AAC Encoder"
 FFmpegOpus="FFmpeg Opus Encoder"
+FFmpegALAC="ALAC (24-bit)"
+FFmpegFLAC="FLAC (16-bit)"
+FFmpegPCM16Bit="PCM (16-bit)"
+FFmpegPCM24Bit="PCM (24-bit)"
+FFmpegPCM32BitFloat="PCM (32-bit float)"
```
Comment (tytan652) [inline plugins/obs-ffmpeg/data/locale/en-US.ini:8]: Even if there is low chance to have non-FFmpeg alternative of those codec as first-party.

Like AAC and Opus, all of them should have "FFmpeg" inside of their names

File: plugins/obs-ffmpeg/obs-ffmpeg-mux.c [hunk 2]
```diff
@@ -313,11 +313,18 @@ static void build_command_line(struct ffmpeg_muxer *stream, struct dstr *cmd,
 		add_video_encoder_params(stream, cmd, vencoder);
 
 	if (num_tracks) {
-		dstr_cat(cmd, "aac ");
+		struct dstr name = {0};
+		dstr_init_copy(&name, obs_encoder_get_codec(aencoders[0]));
+		dstr_cat(&name, " ");
+		// Name needs to be lowercase
+		dstr_to_lower(&name);
+		dstr_cat_dstr(cmd, &name);
```
Comment (tytan652) [inline plugins/obs-ffmpeg/obs-ffmpeg-mux.c:321]: Forcing lower case is not longer needed since #8090, right ?

Comment (derrod) [reply plugins/obs-ffmpeg/obs-ffmpeg-mux.c:321]: Yes, I have split this out into a separate PR #8496 to fix this issue even if this doesn't get merged yet. I'll rebase this PR later.

**Comment by derrod** [comment]
Rebased now that RFC 45 PRs have been merged, still includes #8496 since those changes are also required for this PR.

**Review by RytoEX — COMMENTED** [review]
Seemed fine at a glance.

**Pull request merged**